lib/ecto/adapters/mssql/connection.ex

```diff
@@ -0,0 +1,1644 @@
+      end
+    end
+
+    defp expr(list, sources, query) when is_list(list) do
+      Enum.map_join(list, ", ", &expr(&1, sources, query))
+    end
+
+    defp expr({string, :varchar}, _sources, _query)
+         when is_binary(string) do
+      "'#{escape_string(string)}'"
+    end
+
+    defp expr(string, _sources, _query) when is_binary(string) do
+      if String.contains?(string, @unsafe_query_strings) do
+        len = String.length(string)
+
+        hex =
+          string
+          |> :unicode.characters_to_binary(:utf8, {:utf16, :little})
+          |> Base.encode16(case: :lower)
+
+        "CONVERT(nvarchar(#{len}), 0x#{hex})"
+      else
+        "N'#{escape_string(string)}'"
+      end
+    end
+
+    defp expr(%Decimal{} = decimal, _sources, _query) do
+      Decimal.to_string(decimal, :normal)
+    end
+
+    defp expr(%Tagged{value: binary, type: :binary}, _sources, _query) when is_binary(binary) do
+      hex = Base.encode16(binary, case: :lower)
+      "0x#{hex}"
+    end
+
+    defp expr(%Tagged{value: binary, type: :uuid}, _sources, _query) when is_binary(binary) do
+      case binary do
+        <<_::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96>> ->
+          {:ok, value} = Tds.Types.UUID.dump(binary)
+          value
+
+        any ->
+          any
+      end
+    end
+
+    defp expr(%Tagged{value: other, type: type}, sources, query)
+         when type in [:varchar, :nvarchar] do
+      "CAST(#{expr(other, sources, query)} AS #{column_type(type, [])}(max))"
+    end
+
+    defp expr(%Tagged{value: other, type: type}, sources, query) do
+      "CAST(#{expr(other, sources, query)} AS #{ecto_cast_to_db(type, [])})"
+    end
+
+    defp expr(nil, _sources, _query), do: "NULL"
+    defp expr(true, _sources, _query), do: "1"
+    defp expr(false, _sources, _query), do: "0"
+
+    defp expr(literal, _sources, _query) when is_binary(literal) do
+      "'#{escape_string(literal)}'"
+    end
+
+    defp expr(literal, _sources, _query) when is_integer(literal) do
+      Integer.to_string(literal)
+    end
+
+    defp expr(literal, _sources, _query) when is_float(literal) do
+      Float.to_string(literal)
+    end
+
+    defp expr(field, sources, query) do
+      error!(
+        query,
+        "MSSQL adapter does not support keyword or interpolated whatever \n#{
+          inspect(inspect([field: field, sources: sources, query: query], structs: false))
+        }"
+      )
+    end
+
+    defp op_to_binary({op, _, [_, _]} = expr, sources, query) when op in @binary_ops do
+      paren_expr(expr, sources, query)
+    end
+
+    defp op_to_binary(expr, sources, query) do
+      expr(expr, sources, query)
+    end
+
+    defp interval_count(count, _sources, _query) when is_integer(count) do
+      String.Chars.Integer.to_string(count)
+    end
+
+    defp interval_count(count, _sources, _query) when is_float(count) do
+      :erlang.float_to_binary(count, [:compact, decimals: 16])
```

Comment:
```suggestion
      Float.to_string(count)
```

Reply:
`Float.to_string` can give unpredictable behavior if there is more than 16 decimal digits. If I recall correctly, token code will be 0x6C and that is not documented in documentation, I'm not sure how to decode that when it gets back to tds driver